### PR TITLE
INTMDB-533: Feature Add: Programmatically Create Organizations

### PR DESCRIPTION
**Pull request by martinstibbe**
## Description

Feature Add: Programmatically Create Organizations

Link to any related issue(s):

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [x] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

### CODE REVIEW

File: website/docs/r/organization.html.markdown [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+layout: "mongodbatlas"
+page_title: "MongoDB Atlas: organization"
+sidebar_current: "docs-mongodbatlas-resource-organization"
+description: |-
+    Provides a Organization resource.
+---
+
+# Resource: mongodbatlas_organization
+
+`mongodbatlas_organization` provides a Organization resource. This allows organization to be created.
```
Comment (Zuhairahmed) [inline website/docs/r/organization.html.markdown]:
```suggestion
`mongodbatlas_organization` provides programmatic management (including creation) of a MongoDB Atlas Organization resource. 
```

File: website/docs/r/organization.html.markdown [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+~> **IMPORTANT NOTE:**  Creating an organization with this resource creates initial public and private key values that will need to be used in a separate terraform module/blueprint to permit access to new organization.
```
Comment (Zuhairahmed) [inline website/docs/r/organization.html.markdown]: Not sure what this means, can you rephrase/make more clear? If easier feel free to borrow language from Atlas Admin API docs - https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Organizations/operation/createOrganization

File: website/docs/r/organization.html.markdown [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  description = "test API key from Org Creation Test"
+  role_names = ["ORG_OWNER"]
+}
+```
+
+## Argument Reference
+
+* `name` - (Required) The name of the organization you want to create. (Cannot be changed via this Provider after creation.)
+* `org_owner_id` - (Required) Unique 24-hexadecimal digit string that identifies the Atlas user that you want to assign the Organization Owner role. This user must be a member of the same organization as the calling API key. This is only required when authenticating with Programmatic API Keys.
+* `description` - API Key description
```
Comment (Zuhairahmed) [inline website/docs/r/organization.html.markdown]:
```suggestion
* `description` - Programmatic API Key description
```

File: website/docs/d/organizations.html.markdown [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+layout: "mongodbatlas"
+page_title: "MongoDB Atlas: organizations"
+sidebar_current: "docs-mongodbatlas-organizations"
+description: |-
+    Describes Organizations.
+---
+
+# Data Source: mongodbatlas_organizations
+
+`mongodbatlas_organizations` describe all Organizations. This represents organizations that have been created.
```
Comment (Zuhairahmed) [inline website/docs/d/organizations.html.markdown]:
```suggestion
`mongodbatlas_organizations` describe all MongoDB Atlas Organizations. This represents organizations that have been created.
```

File: examples/atlas-organization/organization-step-1/main.tf [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  value = mongodbatlas_organization.test.org_id
+}
+
+output "org_public_key" {
+  value = nonsensitive(mongodbatlas_organization.test.public_key)
+}
+
+output "org_private_key" {
+  value = nonsensitive(mongodbatlas_organization.test.private_key)
+}
```
Comment (Zuhairahmed) [inline examples/atlas-organization/organization-step-1/main.tf]:
```suggestion
}

```

File: examples/atlas-organization/organization-step-2/Readme.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+[Amazon Web Services (AWS)](https://docs.atlas.mongodb.com/reference/amazon-aws/#amazon-aws)
+[Google Cloud Platform (GCP)](https://docs.atlas.mongodb.com/reference/google-gcp/#google-gcp)
+[Microsoft Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/#microsoft-azure)
+
+## Dependencies
+
+* Terraform v0.13 or greater
+* A MongoDB Atlas account 
+* provider.mongodbatlas: version = "~> 0.9.1"
+* Some users (see [here](https://github.com/mongodb/terraform-provider-mongodbatlas/issues/1083)) have reported issues deploying this starter example with Mac M1 CPU. If issue persists, suggest deploying instead on x86 linux if possible. See list of supported binaries [here](https://github.com/mongodb/terraform-provider-mongodbatlas/releases/tag/v1.8.1)
```
Comment (Zuhairahmed) [inline examples/atlas-organization/organization-step-2/Readme.md]: I wonder if this is simply user-error of person selecting wrong binary? pretty sure the terraform-provider-mongodbatlas_1.8.1_darwin_arm64.zip binary for example should support Mac M1/M2 machines cc: @maastha @andreaangiolillo

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
+# Example - A basic example to create and Orgnization with MongoDB Atlas and Terraform
```
Comment (Zuhairahmed) [inline examples/atlas-organization/organization-step-1/Readme.md]:
```suggestion
# Example - A basic example to create a MongoDB Atlas Organization with Terraform
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+This project aims to provide a very straight-forward example of setting up and organization Terraform with MongoDB Atlas. This will create the following resources in MongoDB Atlas:
+
+- Atlas organization
+- Private Key
+- Public Key
+- Organization ID
+
+You can refer to the MongoDB Atlas documentation to know about the region names used in MongoDB Atlas respective to the Cloud Provider's region name.
+[Amazon Web Services (AWS)](https://docs.atlas.mongodb.com/reference/amazon-aws/#amazon-aws)
```
Comment (Zuhairahmed) [inline examples/atlas-organization/organization-step-1/Readme.md]:
```suggestion
* [Amazon Web Services (AWS)](https://docs.atlas.mongodb.com/reference/amazon-aws/#amazon-aws)
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+[Google Cloud Platform (GCP)](https://docs.atlas.mongodb.com/reference/google-gcp/#google-gcp)
```
Comment (Zuhairahmed) [inline examples/atlas-organization/organization-step-1/Readme.md]:
```suggestion
* [Google Cloud Platform (GCP)](https://docs.atlas.mongodb.com/reference/google-gcp/#google-gcp)
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+[Microsoft Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/#microsoft-azure)
```
Comment (Zuhairahmed) [inline examples/atlas-organization/organization-step-1/Readme.md]:
```suggestion
* [Microsoft Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/#microsoft-azure)
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
+# Example - A basic example to create and Orgnization with MongoDB Atlas and Terraform
+
+This project aims to provide a very straight-forward example of setting up and organization Terraform with MongoDB Atlas. This will create the following resources in MongoDB Atlas:
+
+- Atlas organization
+- Private Key
+- Public Key
+- Organization ID
+
+You can refer to the MongoDB Atlas documentation to know about the region names used in MongoDB Atlas respective to the Cloud Provider's region name.
```
Comment (Zuhairahmed) [inline examples/atlas-organization/organization-step-1/Readme.md]:
```suggestion
You can refer to the MongoDB Atlas documentation to know about the region names used in MongoDB Atlas respective to the Cloud Provider's region name:
```

File: examples/atlas-organization/organization-step-1/Readme.md (hunk 7 above)
Comment (Zuhairahmed) [inline examples/atlas-organization/organization-step-1/Readme.md]: I think what's missing is a README in atlas-organization folder (directly above) explaining how Step 1 and Step 2 directories relate to one another. Confusing looking at this for the very first time.

**Review by Zuhairahmed — CHANGES_REQUESTED** [review]
Few comments, happy to take another look when ready

**Comment by Zuhairahmed** [comment]
Also @martinstibbe can you resolve the failing GitHub check

File: mongodbatlas/resource_mongodbatlas_organization.go [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+	clients, _ := config.NewClient(ctx)
+	conn := clients.(*MongoDBClient).Atlas
+
+	ids := decodeStateID(d.Id())
+	orgID := ids["org_id"]
+
+	organization, resp, err := conn.Organizations.Get(ctx, orgID)
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound || resp.StatusCode == http.StatusBadRequest {
+			log.Printf("warning Organization deleted will recreate: %s \n", err.Error())
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_organization.go:122]: When do you get `StatusBadRequest` instead of `StatusNotFound`?  if the org is deleted, I would expect to get a NotFound

File: mongodbatlas/resource_mongodbatlas_organization.go [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+
+	clients, _ := config.NewClient(ctx)
+	conn := clients.(*MongoDBClient).Atlas
+	ids := decodeStateID(d.Id())
+	orgID := ids["org_id"]
+
+	_, err := conn.Organizations.Delete(ctx, orgID)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error Organization: %s", err))
+	}
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_organization.go]:
```suggestion
	if _, err := conn.Organizations.Delete(ctx, orgID); err != nil {
		return diag.FromErr(fmt.Errorf("error Organization: %s", err))
	}
```

File: examples/atlas-organization/organization-step-1/Readme.md (hunk 7 above)
Comment (colm-quinn) [inline examples/atlas-organization/organization-step-1/Readme.md]: I think we're missing a README.MD in the root of `atlas-organization` that links together the two steps & explains the overall flow & why it's a 2-step process

Comment (colm-quinn) [reply examples/atlas-organization/organization-step-1/Readme.md]: Also a question for the 2 steps - could we instead have made 2 local modules in the example and then called them from a root module `main.tf` passing parameters between them as necessary? It would save the user needing to do two `terraform apply`, but I may be missing something in the staging of the process here.

Comment (colm-quinn) [reply examples/atlas-organization/organization-step-1/Readme.md:1]: Also a question for the 2 steps - could we instead have made 2 modules in the example and then called them from a root module `main.tf` passing parameters between them as necessary? IT would save the user needing to do two `terraform apply`, but I may be missing something in the staging of the process here.

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
+# Example - A basic example to create and Orgnization with MongoDB Atlas and Terraform
+
+This project aims to provide a very straight-forward example of setting up and organization Terraform with MongoDB Atlas. This will create the following resources in MongoDB Atlas:
```
Comment (themantissa) [inline examples/atlas-organization/organization-step-1/Readme.md]:
```suggestion
This project aims to provide a very straight-forward example of setting up a MongoDB Atlas Organization with Terraform. This will create the following resources in MongoDB Atlas:
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+You can refer to the MongoDB Atlas documentation to know about the region names used in MongoDB Atlas respective to the Cloud Provider's region name.
+[Amazon Web Services (AWS)](https://docs.atlas.mongodb.com/reference/amazon-aws/#amazon-aws)
+[Google Cloud Platform (GCP)](https://docs.atlas.mongodb.com/reference/google-gcp/#google-gcp)
+[Microsoft Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/#microsoft-azure)
+
+## Dependencies
+
+* Terraform v0.13 or greater
+* A MongoDB Atlas account 
+* provider.mongodbatlas: version = "~> 0.9.1"
```
Comment (themantissa) [inline examples/atlas-organization/organization-step-1/Readme.md]:
```suggestion
* provider.mongodbatlas: version = "~> 1.10.0"
* [Cross-organization billing](https://www.mongodb.com/docs/atlas/billing/#cross-organization-billing) enabled and the requesting API Key's organization must be a paying organization. 
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+**2\. Review the Terraform plan.**
+
+Execute the below command and ensure you are happy with the plan.
+
+``` bash
+$ terraform plan
+```
+
+This project currently creates the below deployments:
```
Comment (themantissa) [inline examples/atlas-organization/organization-step-1/Readme.md]:
```suggestion
This example currently creates the following:
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+**4\. Destroy the resources.**
+
+Once you are finished your testing, ensure you destroy the resources to avoid unnecessary charges.
+
+``` bash
+$ terraform destroy
+```
+
+**Important Point**
```
Comment (themantissa) [inline examples/atlas-organization/organization-step-1/Readme.md]: I'll admit this gave me a giggle.  I often have important points, only to forget them.  This is a perfect representation of those times :) 
```suggestion
**Important Point**
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+
+**Output:**
+```
+plstring = mongodb+srv://cluster-atlas-pl-0.za3fb.mongodb.net
+```
```
Comment (themantissa) [inline examples/atlas-organization/organization-step-1/Readme.md]: Not an output.
```suggestion
**Output:**
```
plstring = mongodb+srv://cluster-atlas-pl-0.za3fb.mongodb.net
```
```

File: examples/atlas-organization/organization-step-2/Readme.md [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+You can refer to the MongoDB Atlas documentation to know about the region names used in MongoDB Atlas respective to the Cloud Provider's region name.
+[Amazon Web Services (AWS)](https://docs.atlas.mongodb.com/reference/amazon-aws/#amazon-aws)
+[Google Cloud Platform (GCP)](https://docs.atlas.mongodb.com/reference/google-gcp/#google-gcp)
+[Microsoft Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/#microsoft-azure)
+
+## Dependencies
+
+* Terraform v0.13 or greater
+* A MongoDB Atlas account 
+* provider.mongodbatlas: version = "~> 0.9.1"
```
Comment (themantissa) [inline examples/atlas-organization/organization-step-2/Readme.md]: Doesn't require greater but to keep it consistent with step 1
```suggestion
* provider.mongodbatlas: version = "~> 1.10.0"
```

File: examples/atlas-organization/organization-step-1/Readme.md (hunk 7 above)
Comment (themantissa) [reply examples/atlas-organization/organization-step-1/Readme.md]: +1 on these - we can't assume anyone will figure this out on their own.

File: website/docs/r/organization.html.markdown (hunk 2 above)
Comment (themantissa) [reply website/docs/r/organization.html.markdown]: He's saying that you can't start creating things with this new key pair in the same Terraform apply that creates the new organization.   Just need to rephrase it to be a bit more clear.

File: mongodbatlas/resource_mongodbatlas_organization.go [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+		}
+		return diag.FromErr(fmt.Errorf("error deleting organization information: %s", err))
+	}
+	d.SetId(encodeStateID(map[string]string{
+		"org_id": organization.ID,
+	}))
+	return nil
+}
+
+func resourceMongoDBAtlasOrganizationUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```
Comment (themantissa) [inline mongodbatlas/resource_mongodbatlas_organization.go:134]: Is rename blocked if it's not an API key within the new org (Haven't tried but name is update-able): https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Organizations/operation/renameOrganization

File: website/docs/r/organization.html.markdown [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+  org_owner_id = "6205e5fffff79cde6f"
+  name = "testCreateORG"
+  description = "test API key from Org Creation Test"
+  role_names = ["ORG_OWNER"]
+}
+```
+
+## Argument Reference
+
+* `name` - (Required) The name of the organization you want to create. (Cannot be changed via this Provider after creation.)
```
Comment (themantissa) [inline website/docs/r/organization.html.markdown:29]: This might need to change, depending on the answer to my question in the resource code.

File: website/docs/r/organization.html.markdown [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+* `org_owner_id` - (Required) Unique 24-hexadecimal digit string that identifies the Atlas user that you want to assign the Organization Owner role. This user must be a member of the same organization as the calling API key. This is only required when authenticating with Programmatic API Keys.
```
Comment (themantissa) [inline website/docs/r/organization.html.markdown]:
```suggestion
* `org_owner_id` - (Required) Unique 24-hexadecimal digit string that identifies the Atlas user that you want to assign the Organization Owner role. This user must be a member of the same organization as the calling API key.  This is only required when authenticating with Programmatic API Keys.  [https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/MongoDB-Cloud-Users/operation/getUserByUsername](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/MongoDB-Cloud-Users/operation/getUserByUsername)
```

FYI @Zuhairahmed - another reason we need Atlas Users added soon. Right now you can't use a data source to easily get the Atlas user id within the provider.

File: website/docs/r/organization.html.markdown [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+* `description` - API Key description
+
+~> **NOTE:** Organization created will return a new API Key that would be used by other Terraform modules/blueprints.
```
Comment (themantissa) [inline website/docs/r/organization.html.markdown]: Trying to make a bit more clear:
```suggestion
~> **NOTE:** Creating an organization will return a new API Key pair that can be used to authenticate and manage the new organization  with MongoDB Atlas Terraform modules/blueprints.  You cannot use the newly created API key pair to manage the newly created organization in the same Terraform module/blueprint that the organization is created in.
```

File: website/docs/r/organization.html.markdown [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+  * `ORG_MEMBER` 
+ 
+  
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `org_id` - The organization id.
+* `public_key` - Public API key value set for the specified organization API key.
+* `private_key` - Redacted private key returned for this organization API key. This key displays unredacted when first created.
```
Comment (themantissa) [inline website/docs/r/organization.html.markdown]:
```suggestion
* `private_key` - Redacted private key returned for this organization API key. This key displays unredacted when first created and is saved within the Terraform state file.
```

File: examples/atlas-organization/organization-step-2/Readme.md [hunk 25]
```diff
@@ -0,0 +1,64 @@
+# Example - A basic example to start with the MongoDB Atlas and Terraform
+
+This project aims to provide a very straight-forward example of setting up Terraform with MongoDB Atlas. This will create the following resources in MongoDB Atlas:
```
Comment (themantissa) [inline examples/atlas-organization/organization-step-2/Readme.md:3]: Should we not explain how to get the key pair from step one to step 2?  Again that calls for an overarching readme.

**Review by themantissa — CHANGES_REQUESTED** [review]
So exciting to finally be *this close* to org creation!   A few comments/questions/suggestions.

File: examples/atlas-organization/organization-step-1/Readme.md (hunk 7 above)
Comment (martinstibbe) [reply examples/atlas-organization/organization-step-1/Readme.md]: Yep was trying to keep them isolated so the do not try to use old API keys on new org in next apply needs work I agree

File: website/docs/r/organization.html.markdown (hunk 22 above)
Comment (Zuhairahmed) [reply website/docs/r/organization.html.markdown]: makes sense @themantissa, this should be part of near term future release

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 26]
```diff
@@ -0,0 +1,74 @@
+# Example - A basic example to create and Orgnization with MongoDB Atlas and Terraform
+
+This project aims to provide a very straight-forward example of setting up a MongoDB Atlas Organization with Terraform. This will create the following resources in MongoDB Atlas:
+
+- Atlas organization
+- Private Key
+- Public Key
+- Organization ID
+
+You can refer to the MongoDB Atlas documentation to know about the region names used in MongoDB Atlas respective to the Cloud Provider's region name:
```
Comment (zach-carr) [inline examples/atlas-organization/organization-step-1/Readme.md]: suggest replacing "know" with an action
```suggestion
You can refer to the MongoDB Atlas documentation to learn more about the region names used in MongoDB Atlas respective to the Cloud Provider's region name:
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+* [Google Cloud Platform (GCP)](https://docs.atlas.mongodb.com/reference/google-gcp/#google-gcp)
+* [Microsoft Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/#microsoft-azure)
+
+## Dependencies
+
+* Terraform v0.13 or greater
+* A MongoDB Atlas account 
+* provider.mongodbatlas: version = "~> 1.10.0"
+* [Cross-organization billing](https://www.mongodb.com/docs/atlas/billing/#cross-organization-billing) enabled and the requesting API Key's organization must be a paying organization. 
+* Some users (see [here](https://github.com/mongodb/terraform-provider-mongodbatlas/issues/1083)) have reported issues deploying this starter example with Mac M1 CPU. If issue persists, suggest deploying instead on x86 linux if possible. See list of supported binaries [here](https://github.com/mongodb/terraform-provider-mongodbatlas/releases/tag/v1.8.1)
```
Comment (zach-carr) [inline examples/atlas-organization/organization-step-1/Readme.md]: suggest more directly addressing the user for clarity - "If you encounter this issue"
```suggestion
* Some users (see [here](https://github.com/mongodb/terraform-provider-mongodbatlas/issues/1083)) have reported issues deploying this starter example with Mac M1 CPU. If you encounter this issue, try deploying instead on x86 linux if possible. See a list of supported binaries [here](https://github.com/mongodb/terraform-provider-mongodbatlas/releases/tag/v1.8.1).
```

File: examples/atlas-organization/organization-step-1/Readme.md [hunk 28]
```diff
@@ -0,0 +1,74 @@
...
+**1\. Ensure your MongoDB Atlas credentials are set up.**
+
+This can be done using environment variables:
+
+```bash
+export MONGODB_ATLAS_PUBLIC_KEY="xxxx"
+export MONGODB_ATLAS_PRIVATE_KEY="xxxx"
+```
+
+... or follow as in the `variables.tf` file and create **terraform.tfvars** file with all the variable values and make sure **not to commit it**.
```
Comment (zach-carr) [inline examples/atlas-organization/organization-step-1/Readme.md]: to me, "follow as in" is not totally clear - does this mean follow `variables.tf` as an example?

File: examples/atlas-organization/organization-step-1/variables.tf [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+  type        = string
+  description = "MongoDB Atlas Database User Password"
+}
+variable "database_name" {
+  type        = string
+  description = "The database in the cluster to limit the database user to, the database does not have to exist yet"
+}
+variable "ip_address" {
+  type        = string
+  description = "The IP address that the cluster will be accessed from, can also be a CIDR range or AWS security group"
```
Comment (zach-carr) [inline examples/atlas-organization/organization-step-1/variables.tf]: minor suggestion to make this clearer that it's a security feature and you won't be able to connect from elsewhere
```suggestion
  description = "The IP address that the cluster may be accessed from, can also be a CIDR range or AWS security group"
```

File: examples/atlas-organization/organization-step-2/Readme.md [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+
+- Atlas Project
+
+You can refer to the MongoDB Atlas documentation to know about the region names used in MongoDB Atlas respective to the Cloud Provider's region name.
```
Comment (zach-carr) [inline examples/atlas-organization/organization-step-2/Readme.md]: same suggestion as above
```suggestion
You can refer to the MongoDB Atlas documentation to learn more about the region names used in MongoDB Atlas respective to the Cloud Provider's region name.
```

File: examples/atlas-organization/organization-step-2/Readme.md (hunk 6 above)
Comment (zach-carr) [inline examples/atlas-organization/organization-step-2/Readme.md]: same comment as above
```suggestion
* Some users (see [here](https://github.com/mongodb/terraform-provider-mongodbatlas/issues/1083)) have reported issues deploying this starter example with Mac M1 CPU. If you encounter this issue, try deploying instead on x86 linux if possible. See list of supported binaries [here](https://github.com/mongodb/terraform-provider-mongodbatlas/releases/tag/v1.8.1)  
```

File: mongodbatlas/resource_mongodbatlas_organization.go [hunk 31]
```diff
@@ -0,0 +1,178 @@
...
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			d.SetId("")
+			return nil
+		}
+
+		return diag.FromErr(fmt.Errorf("error create Organization: %s", err))
+	}
+
+	if err := d.Set("private_key", organization.APIKey.PrivateKey); err != nil {
+		return diag.FromErr(fmt.Errorf("error setting `public_key`: %s", err))
```
Comment (zach-carr) [inline mongodbatlas/resource_mongodbatlas_organization.go]: should this be private_key? 
```suggestion
		return diag.FromErr(fmt.Errorf("error setting `private_key`: %s", err))
```

File: website/docs/d/organizations.html.markdown [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+
+## Argument Reference
+* `page_num` - (Optional)  	The page to return. Defaults to `1`.
+* `items_per_page` - (Optional) Number of items to return per page, up to a maximum of 500. Defaults to `100`.
+
+
+* `id` - Autogenerated Unique ID for this data source.
+* `total_count` - Represents the total of the organizations
+
+* `results` - A list where each represents an Organization.
```
Comment (zach-carr) [inline website/docs/d/organizations.html.markdown]: it's intuitive as-is, but I think "item" or "object" is necessary 
```suggestion
* `results` - A list where each item represents an Organization.
```

File: website/docs/d/organizations.html.markdown [hunk 33]
```diff
@@ -0,0 +1,40 @@
...
+}
+```
+
+## Argument Reference
+* `page_num` - (Optional)  	The page to return. Defaults to `1`.
+* `items_per_page` - (Optional) Number of items to return per page, up to a maximum of 500. Defaults to `100`.
+
+
+* `id` - Autogenerated Unique ID for this data source.
+* `total_count` - Represents the total of the organizations
```
Comment (zach-carr) [inline website/docs/d/organizations.html.markdown]: I think this could be clearer
```suggestion
* `total_count` - Represents the total number of organizations
```

**Review by zach-carr — COMMENTED** [review]
Left several minor comments, thanks

File: mongodbatlas/resource_mongodbatlas_organization.go (hunk 12 above)
Comment (martinstibbe) [reply mongodbatlas/resource_mongodbatlas_organization.go:122]: Corrected

File: mongodbatlas/resource_mongodbatlas_organization.go (hunk 31 above)
Comment (martinstibbe) [reply mongodbatlas/resource_mongodbatlas_organization.go]: corrected

File: website/docs/d/organizations.html.markdown (hunk 33 above)
Comment (martinstibbe) [reply website/docs/d/organizations.html.markdown]: corrected

File: website/docs/d/organizations.html.markdown (hunk 32 above)
Comment (martinstibbe) [reply website/docs/d/organizations.html.markdown]: corrected

File: examples/atlas-organization/organization-step-2/Readme.md (hunk 6 above)
Comment (martinstibbe) [reply examples/atlas-organization/organization-step-2/Readme.md]: corrected

File: examples/atlas-organization/organization-step-2/Readme.md (hunk 30 above)
Comment (martinstibbe) [reply examples/atlas-organization/organization-step-2/Readme.md]: Removed region statement as not relevant in this example

File: examples/atlas-organization/organization-step-1/variables.tf (hunk 29 above)
Comment (martinstibbe) [reply examples/atlas-organization/organization-step-1/variables.tf]: Removed IP address  variable as not relevant in this example

File: examples/atlas-organization/organization-step-1/Readme.md (hunk 28 above)
Comment (martinstibbe) [reply examples/atlas-organization/organization-step-1/Readme.md]: corrected to say utiliize

File: examples/atlas-organization/organization-step-1/Readme.md (hunk 27 above)
Comment (martinstibbe) [reply examples/atlas-organization/organization-step-1/Readme.md]: corrected

File: examples/atlas-organization/organization-step-1/Readme.md (hunk 26 above)
Comment (martinstibbe) [reply examples/atlas-organization/organization-step-1/Readme.md]: Removed region statement as not relevant in this example

File: examples/atlas-organization/organization-step-1/Readme.md (hunk 28 above)
Comment (martinstibbe) [reply examples/atlas-organization/organization-step-1/Readme.md]: changed to utilize variables.tf

**Review by andreaangiolillo — APPROVED** [review]
LGTM

File: examples/atlas-organization/Readme.md [hunk 34]
```diff
@@ -0,0 +1,155 @@
+
+# Example - A basic example to create and Orgnization with MongoDB Atlas and Terraform
+
+This project aims to provide a very straight-forward example of setting up a MongoDB Atlas Organization with Terraform. This will create the following resources in MongoDB Atlas:
+
+- Atlas organization
```
Comment (Zuhairahmed) [inline examples/atlas-organization/Readme.md]:
```suggestion
- MongoDB Atlas organization
```

File: examples/atlas-organization/Readme.md [hunk 35]
```diff
@@ -0,0 +1,155 @@
...
+- Private Key
+- Public Key
+- Organization ID
+- Atlas Project
```
Comment (Zuhairahmed) [inline examples/atlas-organization/Readme.md]:
```suggestion
- MongoDB Atlas Project
```

**Review by Zuhairahmed — APPROVED** [review]
LGTM pending doc feedbacks from my comments earlier

File: examples/atlas-organization/organization-step-1/Readme.md (hunk 7 above)
Comment (themantissa) [reply examples/atlas-organization/organization-step-1/Readme.md]: Resolved for me, @colm-quinn leaving open as it's your point.

File: examples/atlas-organization/organization-step-2/Readme.md (hunk 19 above)
Comment (themantissa) [reply examples/atlas-organization/organization-step-2/Readme.md]: Not going to block on this but would be good to get updated before release or soon there after.

File: website/docs/r/organization.html.markdown (hunk 2 above)
Comment (themantissa) [reply website/docs/r/organization.html.markdown]: Still could use improvement but not blocking on it.

File: mongodbatlas/resource_mongodbatlas_organization.go (hunk 20 above)
Comment (themantissa) [reply mongodbatlas/resource_mongodbatlas_organization.go:134]: Still an open question, @andreaangiolillo or @colm-quinn perhaps you can assist?

File: website/docs/r/organization.html.markdown (hunk 21 above)
Comment (themantissa) [reply website/docs/r/organization.html.markdown:29]: Still need an answer on https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1176/files#r1213704086

**Review by themantissa — CHANGES_REQUESTED** [review]
A few things that are "nice to have" but not must have.  This one however: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1176/files#r1213704086 I'd like to get an answer on before I approve.   Thanks!!

File: mongodbatlas/resource_mongodbatlas_organization.go (hunk 20 above)
Comment (martinstibbe) [reply mongodbatlas/resource_mongodbatlas_organization.go:134]: At the moment there is no update API in TF side as SDK did not have update support

File: website/docs/r/organization.html.markdown [hunk 36]
```diff
@@ -0,0 +1,61 @@
...
+sidebar_current: "docs-mongodbatlas-resource-organization"
+description: |-
+    Provides a Organization resource.
+---
+
+# Resource: mongodbatlas_organization
+
+`mongodbatlas_organization` provides programmatic management (including creation) of a MongoDB Atlas Organization resource.
+
+~> **IMPORTANT NOTE:**  Creating an organization with this resource creates initial public and private key values that will need to be used in a separate terraform module/blueprint to permit access to new organization.
```
Comment (Zuhairahmed) [inline website/docs/r/organization.html.markdown]:
```suggestion
~> **IMPORTANT NOTE:**  When you establish an Atlas organization using this resource, it automatically generates a set of initial public and private Programmatic API Keys. These key values are vital to store because you'll need to use them to grant access to the newly created Atlas organization.
```

File: website/docs/r/organization.html.markdown (hunk 2 above)
Comment (Zuhairahmed) [reply website/docs/r/organization.html.markdown]: thanks. updated directly in case you want to take another look @themantissa

File: mongodbatlas/resource_mongodbatlas_organization.go (hunk 20 above)
Comment (Zuhairahmed) [reply mongodbatlas/resource_mongodbatlas_organization.go:134]: Given users can rename via Atlas Admin API suspect they are going to push us to allow them to also do this in Terraform Provider as well. Suggest if no concerns we update SDK then flow through to here accordingly.

Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_organization.go:134]: no concern, I will update the SDK today and add the missing endpoints. I will add an update here as soon as the endpoint is available

Comment (andreaangiolillo) [reply mongodbatlas/resource_mongodbatlas_organization.go:134]: the update endpoint is available to the SDK. You can download the latest version with go get [go.mongodb.org/atlas@INTMDB-860](http://go.mongodb.org/atlas@INTMDB-860) . I will release the SDK later today so that we have a new version before the TF release.

Comment (Zuhairahmed) [reply mongodbatlas/resource_mongodbatlas_organization.go:134]: @themantissa as an update rename endpoint has been added to this PR. thanks again @martinstibbe  + @andreaangiolillo for rapid speed here.

File: website/docs/r/organization.html.markdown (hunk 21 above)
Comment (Zuhairahmed) [reply website/docs/r/organization.html.markdown:29]: @martinstibbe i think we are missing `isDeleted` here: https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Organizations/operation/renameOrganization

Comment (martinstibbe) [reply website/docs/r/organization.html.markdown:29]: Added isDeleted

File: mongodbatlas/resource_mongodbatlas_organization.go (hunk 20 above)
Comment (themantissa) [reply mongodbatlas/resource_mongodbatlas_organization.go:134]: Speeeeeeedy!  Thanks all!

**Review by themantissa — APPROVED** [review]
Amazing turn around time on the update improvement.  LGTM!
